File: LeetCode/remote/binary_tree_level_order_traversal_II.rb
#   https://leetcode.com/problems/binary-tree-level-order-traversal-ii/
#   Given a binary tree, return the bottom-up level order traversal of its nodes' values.
#   (ie, from left to right, level by level from leaf to root).
#
#   https://leetcode.com/submissions/detail/54734518/
#
#   Submission Details
#   34 / 34 test cases passed.
#       Status: Accepted
#       Runtime: 78 ms
#           
#           Submitted: 0 minutes ago
#   You are here!
#   Your runtime beats 100.00% of rubysubmissions.

def level_order_bottom(root)
    result = []    
    level_order(root, result, 0)
    result.reverse
end

def level_order(root, result, level)
    return if root.nil?
    result[level] ||= []
    result[level] << root.val
    level_order(root.left, result, level + 1)
    level_order(root.right, result, level + 1)
end

File: LeetCode/remote/binary_tree_level_order_traversal_II.cs
//  https://leetcode.com/problems/binary-tree-level-order-traversal-ii/
//  Given a binary tree, return the bottom-up level order traversal of its nodes' values. (ie, from left to right, level by level from leaf to root).
//
//  https://leetcode.com/submissions/detail/60352465/
//
//  Submission Details
//  34 / 34 test cases passed.
//      Status: Accepted
//      Runtime: 680 ms
//          
//          Submitted: 0 minutes ago

public class Solution {
    public IList<IList<int>> LevelOrderBottom(TreeNode root) {
        var result = new List<IList<int>>();
        if (root == null)
        {
            return result;
        }
        
        var left = LevelOrderBottom(root.left);
        var right = LevelOrderBottom(root.right);
        var i = left.Count;
        var j = right.Count;
        while (i > 0 || j > 0)
        {
            result.Add(
                (i > 0 ? left[--i] : new List<int>())
                .Concat(j > 0 ? right[--j] : new List<int>())
                .ToList());
        }

        result.Reverse();
        result.Add(new List<int> { root.val });
        return result;
    }
}
